### PR TITLE
hexaddress decoder test + json double encoding fix

### PR DESCRIPTION
**Pull request by hackaugusto**

### CODE REVIEW

File: raiden/api/v1/encoding.py [hunk 1]
```diff
@@ -35,16 +35,15 @@
 )
 
 
-# type converter for the flask route
 class HexAddressConverter(BaseConverter):
-
     def to_python(self, value):
-        value = address_decoder(value)
-        return value
+        if value[:2] == '0x':
```
Comment (LefterisJP) [inline raiden/api/v1/encoding.py]: This makes our input to the API endpoint stricter. Before by using the `address_decoder()` we were not enforcing the `0x` prefix. We accepted hex addresses both with and without it. Do we really want to enforce this?

File: raiden/api/rest.py [hunk 2]
```diff
@@ -339,37 +343,40 @@ def patch_channel(self, channel_address, balance=None, state=None):
...
-            else:
-                # should never happen, channel_state is validated in the schema
+            raiden_service_result = self.raiden_api.close(
+                channel.token_address,
+                channel.partner_address
+            )
+            result = self.channel_schema.dump(channel_to_api_dict(raiden_service_result))
+            return jsonify(result.data)
+
+        if state == CHANNEL_STATE_SETTLED:
```
Comment (LefterisJP) [inline raiden/api/rest.py]: why not `elif` here?

File: raiden/api/v1/encoding.py (hunk 1 above)
Comment (hackaugusto) [inline raiden/api/v1/encoding.py]: I have no strong argument in favor of it other than it's my preference to be stricter, relaxing rules is backwards compatible the opposite is not (e.g. geth's change in the address encoding for the genesis block)

File: raiden/api/rest.py (hunk 2 above)
Comment (hackaugusto) [inline raiden/api/rest.py]: Because the previous block is returning

Comment (LefterisJP) [inline raiden/api/rest.py]: but still it's visually easier to follow the code if there is an elif there

File: raiden/api/v1/encoding.py (hunk 1 above)
Comment (LefterisJP) [inline raiden/api/v1/encoding.py]: Okay as discussed between the team we are a go with changing this but the change here will not work. In fact we should remove this HexAdddressConverter as it is not used at all.

Please remove it from api/rest.py and all the other places where it is located.

File: raiden/api/v1/encoding.py [hunk 3]
```diff
@@ -35,16 +35,15 @@
...
+
+        raise Exception('invalid address')
 
     def to_url(self, value):
-        value = address_encoder(value)
-        return value
+        return address_encoder(value)
 
 
 class AddressField(fields.Field):
```
Comment (LefterisJP) [inline raiden/api/v1/encoding.py]: Instead @hackaugusto you need to put the logic here in the `_deserialize()` method of the `AddressField`

Also please make a test in `test_api.py` where addresses without the `0x` prefix are given as input and check that we get back an appropriate error from the API.

I did some tests and if you throw from the `_deserialize()` method we get a simple `500 Internal Server Error` return and no more details. That is not sufficient as we need to alert the user that he needs to add the 0x prefix.

The solution is to throw a specific exception and then add an error handler for that exception to the flask app as detailed [here](http://flask.pocoo.org/docs/0.12/patterns/apierrors/)

This is not gonna work for us and should find a way to alert the user that he gave invalid input.

Comment (hackaugusto) [inline raiden/api/v1/encoding.py]: done

Comment (LefterisJP) [inline raiden/api/v1/encoding.py]: thanks. Just 2 more smaller comments left. Looks nice now :)

File: raiden/tests/api/test_api.py [hunk 4]
```diff
@@ -80,6 +105,71 @@ def test_channel_to_api_dict(raiden_network, token_addresses, settle_timeout):
...
+        'channels/ea674fdde714fd979de3edf0f56aa9716b898ec8'
+    ).format(port=rest_api_port_number)
+
+    request = grequests.patch(
+        url_without_prefix,
+        json={'state': CHANNEL_STATE_SETTLED}
+    )
+    response = request.send().response
+
+    assert response.status_code == httplib.NOT_FOUND
```
Comment (hackaugusto) [inline raiden/tests/api/test_api.py]: URL is universal resource locator, a bad URL means the resource does not exist, also this is controlled by flask.

Comment (LefterisJP) [inline raiden/tests/api/test_api.py]: I am familiar with what a URL is. I am asking if there is a way to make flask return an error here indicating that the reason the resource was not found is that the prefix is missing.

Essentially if there is any way to expose `HexAddressConverter` errors and show them to the user.

**Comment by LefterisJP** [comment]
All ready to go! Merging.

**Pull request merged**